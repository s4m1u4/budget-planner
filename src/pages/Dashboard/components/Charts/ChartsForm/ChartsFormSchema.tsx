import * as yup from "yup";

export const ChartsFormSchema = yup.object({
  title: yup.string().required("Title is required"),
  description: yup.string().required("Description is required"),
});
