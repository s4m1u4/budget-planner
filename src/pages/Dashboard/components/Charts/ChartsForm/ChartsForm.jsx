import React from "react";
import { ChartsFormSchema } from "./ChartsFormSchema";
import { useFormik } from "formik";
import { ModalComponent } from "../../../../../components/shared";
import { ButtonGroup, Form, Title } from "./ChartsForm.styles";
import { Button, TextField } from "@mui/material";

export const ChartsForm = ({ open, handleClose, setNewCategory }) => {
  const handleSubmit = async (values) => {
    await setNewCategory(values);
  };

  const formik = useFormik({
    initialValues: {
      title: "",
      description: "",
    },
    validationSchema: ChartsFormSchema,
    onSubmit: async (values) => {
      await handleSubmit(values);
      formik.resetForm();
      handleClose();
    },
  });

  return (
    <ModalComponent open={open} onClose={handleClose}>
      <Title>Create a new category</Title>
      <Form onSubmit={formik.handleSubmit}>
        <TextField
          fullWidth
          name="title"
          size="small"
          label="Title"
          value={formik.values.title}
          onBlur={formik.handleBlur}
          onChange={formik.handleChange}
          error={formik.touched.title && !!formik.errors.title}
          helperText={formik.touched.title && formik.errors.title}
        />
        <TextField
          fullWidth
          name="description"
          size="small"
          label="Description"
          value={formik.values.description}
          onBlur={formik.handleBlur}
          onChange={formik.handleChange}
          error={formik.touched.description && !!formik.errors.description}
          helperText={formik.touched.description && formik.errors.description}
        />
        <ButtonGroup>
          <Button variant="contained" color="success" type="submit">
            Add category
          </Button>
          <Button variant="contained" color="error" onClick={handleClose}>
            Close
          </Button>
        </ButtonGroup>
      </Form>
    </ModalComponent>
  );
};
