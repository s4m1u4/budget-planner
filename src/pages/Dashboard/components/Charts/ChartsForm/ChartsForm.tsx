import React, { FC } from "react";
import { useFormik } from "formik";
import { ChartsFormSchema } from "./ChartsFormSchema";
import { ModalComponent } from "../../../../../components/shared";
import { Button, TextField } from "@mui/material";
import { ICategoryData } from "../../../types";

import { ButtonGroup, Form, Title } from "./ChartsForm.styles";

interface ChartsFormProps {
  open: boolean;
  onSubmit: (values: ICategoryData) => void;
  handleClose: () => void;
  setNewCategory: (categoryData: ICategoryData) => void;
}

export const ChartsForm: FC<ChartsFormProps> = ({
  open,
  onSubmit,
  handleClose,
  setNewCategory,
}) => {
  const handleSubmit = async (values: ICategoryData) => {
    await setNewCategory(values);
  };

  const formik = useFormik({
    initialValues: {
      title: "",
      description: "",
    },
    validationSchema: ChartsFormSchema,
    onSubmit: async (values: ICategoryData) => {
      await handleSubmit(values);
      await onSubmit(values);
      formik.resetForm();
      handleClose();
    },
  });

  return (
    <ModalComponent open={open} onClose={handleClose}>
      <Title>Create a new category</Title>
      <Form onSubmit={formik.handleSubmit}>
        <TextField
          fullWidth
          name="title"
          size="small"
          label="Title"
          value={formik.values.title}
          onBlur={formik.handleBlur}
          onChange={formik.handleChange}
          error={formik.touched.title && !!formik.errors.title}
          helperText={formik.touched.title && formik.errors.title}
        />
        <TextField
          fullWidth
          name="description"
          size="small"
          label="Description"
          value={formik.values.description}
          onBlur={formik.handleBlur}
          onChange={formik.handleChange}
          error={formik.touched.description && !!formik.errors.description}
          helperText={formik.touched.description && formik.errors.description}
        />
        <ButtonGroup>
          <Button
            sx={{ color: "#fff" }}
            variant="contained"
            color="success"
            type="submit"
          >
            Add category
          </Button>
          <Button
            sx={{ color: "#fff" }}
            variant="contained"
            color="error"
            onClick={handleClose}
          >
            Close
          </Button>
        </ButtonGroup>
      </Form>
    </ModalComponent>
  );
};
