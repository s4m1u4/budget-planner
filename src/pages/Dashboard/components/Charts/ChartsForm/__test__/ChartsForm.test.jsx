import { ChartsForm } from "../ChartsForm";
import { act, render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

const renderComponent = () => {
  return render(
    <ChartsForm
      open={open}
      onSubmit={onSubmit}
      handleClose={handleClose}
      setNewCategory={setNewCategory}
    />
  );
};

const renderComponentAsync = async () => {
  await act(
    async () =>
      await render(
        <ChartsForm
          open={open}
          onSubmit={onSubmit}
          handleClose={handleClose}
          setNewCategory={setNewCategory}
        />
      )
  );
};

const open = true;
const onSubmit = jest.fn();
const handleClose = jest.fn();
const setNewCategory = jest.fn();

describe("ChartsForm component", () => {
  it("snapshot matching", () => {
    const { baseElement } = renderComponent();
    expect(baseElement).toMatchSnapshot();
  });

  describe("should render", () => {
    beforeEach(() => {
      renderComponent();
    });

    it("the title", () => {
      expect(
        screen.getByRole("heading", { name: /create a new category/i })
      ).toBeInTheDocument();
    });

    it('an input with name="title"', () => {
      expect(screen.getByLabelText(/title/i)).toBeInTheDocument();
    });

    it('an input with name="description"', () => {
      expect(screen.getByLabelText(/description/i)).toBeInTheDocument();
    });

    it('a button "Add record"', () => {
      expect(
        screen.getByRole("button", { name: /add category/i })
      ).toBeInTheDocument();
    });

    it('a button "Close"', () => {
      expect(
        screen.getByRole("button", { name: /close/i })
      ).toBeInTheDocument();
    });

    it("error message when submitting a form with empty inputs", async () => {
      expect(screen.queryByText(/title is required/i)).toBeNull();
      expect(screen.queryByText(/description is required/i)).toBeNull();

      userEvent.click(screen.getByRole("button", { name: /add category/i }));

      await renderComponentAsync();

      expect(screen.getByText(/title is required/i)).toBeInTheDocument();
      expect(screen.getByText(/description is required/i)).toBeInTheDocument();
    });
  });

  it("successful form submission", async () => {
    await renderComponentAsync();

    userEvent.type(screen.getByLabelText(/title/i), "Work");
    userEvent.type(
      screen.getByLabelText(/description/i),
      "Create new category"
    );

    userEvent.click(screen.getByRole("button", { name: /add category/i }));

    await waitFor(() =>
      expect(onSubmit).toHaveBeenCalledWith({
        title: "Work",
        description: "Create new category",
      })
    );

    expect(onSubmit).toHaveBeenCalledTimes(1);
  });
});
