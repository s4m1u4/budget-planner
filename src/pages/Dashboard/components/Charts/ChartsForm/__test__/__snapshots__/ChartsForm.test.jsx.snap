// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartsForm component snapshot matching 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiModal-root css-79ws1d-MuiModal-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-i9fmh8-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="css-5fyqun"
      tabindex="-1"
    >
      <h3
        class="css-ngqp4j"
      >
        Create a new category
      </h3>
      <form
        class="css-ezycoh"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="mui-1"
            id="mui-1-label"
          >
            Title
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
              id="mui-1"
              name="title"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-173wfxe"
              >
                <span>
                  Title
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="mui-2"
            id="mui-2-label"
          >
            Description
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
              id="mui-2"
              name="description"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-173wfxe"
              >
                <span>
                  Description
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="css-20ow0d"
        >
          <button
            class="MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-ke5b6m-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="submit"
          >
            Add category
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-1sszw6i-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Close
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </form>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
